tools/btmidi-server.c

```diff
@@ -0,0 +1,1032 @@
+	btsec.level = sec;
+	if (setsockopt(sk, SOL_BLUETOOTH, BT_SECURITY, &btsec,
+							sizeof(btsec)) != 0) {
+		fprintf(stderr, "Failed to set L2CAP security level\n");
+		goto fail;
+	}
+
+	if (listen(sk, 10) < 0) {
+		perror("Listening on socket failed");
+		goto fail;
+	}
+
+	if (verbose) {
+		printf("Started listening on ATT channel. Waiting for connections\n");
+	}
+
+	memset(&addr, 0, sizeof(addr));
+	optlen = sizeof(addr);
+	nsk = accept(sk, (struct sockaddr *) &addr, &optlen);
+	if (nsk < 0) {
+		perror("Accept failed");
+		goto fail;
+	}
+
+	if (verbose) {
+		ba2str(&addr.l2_bdaddr, ba);
+		printf("Connect from %s\n", ba);
+	}
+	
+	close(sk);
+
+	return nsk;
+
+fail:
+	close(sk);
+	return -1;
+}
+
+static struct server *server_create(int fd, uint16_t mtu, int client, int port)
+{
+	struct server *server;
+	size_t name_len = strlen(gattName);
+	int err;
+
+	server = new0(struct server, 1);
+	if (!server) {
+		fprintf(stderr, "Failed to allocate memory for server\n");
+		return NULL;
+	}
+
+	/* Init values */
+	server->io = NULL;
+	server->seq_port_id = -1;
+	server->seq_handle = NULL;
+
+	server->att = bt_att_new(fd, false);
+	if (!server->att) {
+		fprintf(stderr, "Failed to initialze ATT transport layer\n");
+		goto fail;
+	}
+
+	if (!bt_att_set_close_on_unref(server->att, true)) {
+		fprintf(stderr, "Failed to set up ATT transport layer\n");
+		goto fail;
+	}
+
+	if (!bt_att_register_disconnect(server->att, att_disconnect_cb, server, NULL)) {
+		fprintf(stderr, "Failed to set ATT disconnect handler\n");
+		goto fail;
+	}
+
+#if ENABLE_EXTRA_CHARACTERISTICS
+	server->name_len = name_len + 1;
+	server->device_name = malloc(name_len + 1);
+	if (!server->device_name) {
+		fprintf(stderr, "Failed to allocate memory for device name\n");
+		goto fail;
+	}
+
+	memcpy(server->device_name, gattName, name_len);
+	server->device_name[name_len] = '\0';
+#endif
+
+	server->fd = fd;
+	server->db = gatt_db_new();
+	if (!server->db) {
+		fprintf(stderr, "Failed to create GATT database\n");
+		goto fail;
+	}
+
+	server->gatt = bt_gatt_server_new(server->db, server->att, mtu);
+	if (!server->gatt) {
+		fprintf(stderr, "Failed to create GATT server\n");
+		goto fail;
+	}
+
+	if (verbose) {
+		bt_att_set_debug(server->att, att_debug_cb, "att: ", NULL);
+		bt_gatt_server_set_debug(server->gatt, gatt_debug_cb, "server: ", NULL);
+	}
+
+	/* Populate our databases */
+	populate_gap_service(server);
+	populate_gatt_service(server);
+	err = populate_midi_service(server, client, port);
+	if (err < 0) {
+		fprintf(stderr, "Failed to populate midi service\n");
+		goto fail;
+	}
+
+	return server;
+
+fail:
+	gatt_db_unref(server->db);
+#if ENABLE_EXTRA_CHARACTERISTICS
+	free(server->device_name);
+#endif
+	bt_att_unref(server->att);
+	free(server);
+
+	return NULL;
+}
+
+static void server_destroy(struct server *server)
+{
+	bt_gatt_server_unref(server->gatt);
+	gatt_db_unref(server->db);
+}
+
+static int wait_eligible_port(int *out_client, int *out_port)
```

Comment:
@ftonello My concern is this tool be easy to use. OK, I'll remove this and solve the question with a simple printf("Please, connect something to the ALSA sequencer port BLE-MIDI Device")

@cookj273 you are right: at the moment this code uses the first device. Once this functionality will be remove, don't worry